pgxn/neon/pagestore_smgr.c

```diff
@@ -904,13 +928,44 @@ static NeonResponse *
 page_server_request(void const *req)
 {
 	NeonResponse *resp;
+	BufferTag tag = {0};
+	shardno_t shard_no;
+
+	switch (((NeonRequest *) req)->tag)
+	{
+		case T_NeonExistsRequest:
+			CopyNRelFileInfoToBufTag(tag, ((NeonExistsRequest *) req)->rinfo);
+			break;
+		case T_NeonNblocksRequest:
+			CopyNRelFileInfoToBufTag(tag, ((NeonNblocksRequest *) req)->rinfo);
+			break;
+		case T_NeonDbSizeRequest:
+			NInfoGetDbOid(BufTagGetNRelFileInfo(tag)) = ((NeonDbSizeRequest *) req)->dbNode;
+			break;
+		case T_NeonGetPageRequest:
+			CopyNRelFileInfoToBufTag(tag, ((NeonGetPageRequest *) req)->rinfo);
+			tag.blockNum = ((NeonGetPageRequest *) req)->blkno;
+			break;
+		default:
+			neon_log(ERROR, "Unexpected request tag: %d", ((NeonRequest *) req)->tag);
+	}
+	shard_no = get_shard_number(&tag);
+
+
+	/*
+	 * TODO: temporary workarround - we stream all WAL only to shard 0, so metadata and forks other than main
```

Comment:
Do we have an issue tracking this? Maybe let's add issue number to this comment

Reply:
It is not an issue. It is limitation of current sharding schema.
We do not find good way to broadcast/scatter metadata between shards.
So current decision is that metadata is available only at shard 0.

Reply:
So it's not just a temporary workaround? Please update the comment.

Reply:
Well, there is Russian proverb: "there is nothing more durable than temporary solutions".
Such asymmetry in distributing data between shards seems to be strange and unnatural. 
From the other hand it doesn't violate any goals of using sharding: scaling, reducing startup time,...
Shard zero doesn't become bottleneck  (it is required for normal work as well as any other shard).
So I suspect that we will live with this model long enough.
I will update comment.